homeassistant/components/recorder/history.py

```diff
@@ -240,50 +240,67 @@ def _get_states_with_session(
-        and_(
-            States.entity_id == most_recent_states_by_date.c.max_entity_id,
-            States.last_updated == most_recent_states_by_date.c.max_last_updated,
-        ),
-    )
-
-    most_recent_state_ids = most_recent_state_ids.group_by(States.entity_id)
-
-    most_recent_state_ids = most_recent_state_ids.subquery()
-
-    query = query.join(
-        most_recent_state_ids,
-        States.state_id == most_recent_state_ids.c.max_state_id,
-    )
-
-    if entity_ids is not None:
-        query = query.filter(States.entity_id.in_(entity_ids))
+        # We got an include-list of entities, accelerate the query by filtering already
+        # in the inner query.
+        most_recent_state_ids = (
+            session.query(
+                func.max(States.state_id).label("max_state_id"),
+            )
+            .filter(
+                (States.last_updated >= run.start)
+                & (States.last_updated < utc_point_in_time)
+            )
+            .filter(States.entity_id.in_(entity_ids))
+        )
+        most_recent_state_ids = most_recent_state_ids.group_by(States.entity_id)
+        most_recent_state_ids = most_recent_state_ids.subquery()
+        query = query.join(
+            most_recent_state_ids,
+            States.state_id == most_recent_state_ids.c.max_state_id,
+        )
     else:
+        # We did not get an include-list of entities, query all states in the inner
+        # query, then filter out unwanted domains as well as applying the custom filter.
+        # This filtering can't be done in the inner query because the domain column is
+        # not indexed and we can't control what's in the custom filter.
+        most_recent_states_by_date = (
+            session.query(
+                States.entity_id.label("max_entity_id"),
+                func.max(States.last_updated).label("max_last_updated"),
+            )
+            .filter(
+                (States.last_updated >= run.start)
+                & (States.last_updated < utc_point_in_time)
+            )
+            .group_by(States.entity_id)
+            .subquery()
+        )
+        most_recent_state_ids = (
+            session.query(func.max(States.state_id).label("max_state_id"))
+            .join(
+                most_recent_states_by_date,
+                and_(
+                    States.entity_id == most_recent_states_by_date.c.max_entity_id,
+                    States.last_updated
+                    == most_recent_states_by_date.c.max_last_updated,
+                ),
+            )
+            .group_by(States.entity_id)
+            .subquery()
+        )
+        query = query.join(
+            most_recent_state_ids,
+            States.state_id == most_recent_state_ids.c.max_state_id,
+        )
         query = query.filter(~States.domain.in_(IGNORE_DOMAINS))
-        if filters:
-            query = filters.apply(query)
+
+    if filters:
+        query = filters.apply(query)
```

Comment:
```suggestion
        if filters:
            query = filters.apply(query)
```

Can we avoid applying the filter if we are asking for specific entity ids?

If its the only way users have to filter out which entities get stats, then we might need to keep it.

Reply:
This change was not intentional, fixed in b56148a